src/conduits/RwaInputConduit3.t.sol

```diff
@@ -0,0 +1,316 @@
+        gemA.deny(me);
+
+        pip = new DSValue();
+        pip.poke(bytes32(uint256(1 ether))); // Spot = $1
+
+        spot.file(ilk, bytes32("pip"), address(pip));
+        spot.file(ilk, bytes32("mat"), ray(1 ether));
+        spot.poke(ilk);
+
+        vat.file(ilk, "line", rad(1000 ether));
+        vat.file("Line",      rad(1000 ether));
+    }
+
+    function setUp() public {
+        setUpMCDandPSM();
+
+        testUrn = new TestUrn();
+        inputConduit = new RwaInputConduit3(address(dai), address(usdx), address(psmA), address(testUrn));
+        inputConduit.mate(me);
+    }
+
+    function testRevertOnDeployConduitWithWrongGem() public {
+        vm.expectRevert("RwaInputConduit3/wrong-gem-for-psm");
+        new RwaInputConduit3(address(dai), address(0), address(psmA), address(testUrn));
+    }
+
+    function testRevertOnDeployConduitWithWrongDai() public {
+        vm.expectRevert("RwaInputConduit3/wrong-dai-for-psm");
+        new RwaInputConduit3(address(0), address(usdx), address(psmA), address(testUrn));
+    }
+
+    function testSetWardAndEmitRelyOnDeploy() public {
+        vm.expectEmit(true, false, false, false);
+        emit Rely(address(this));
+
+        RwaInputConduit3 c = new RwaInputConduit3(address(dai), address(usdx), address(psmA), address(testUrn));
+
+        assertEq(c.wards(address(this)), 1);
+    }
+
+    function testCanRelyDeny() public {
+        assertEq(inputConduit.wards(address(0)), 0);
+
+        vm.expectEmit(true, false, false, false);
+        emit Rely(address(0));
+
+        inputConduit.rely(address(0));
+
+        assertEq(inputConduit.wards(address(0)), 1);
+
+        vm.expectEmit(true, false, false, false);
+        emit Deny(address(0));
+
+        inputConduit.deny(address(0));
+
+        assertEq(inputConduit.wards(address(0)), 0);
+    }
+
+    function testCanMateHate() public {
+        assertEq(inputConduit.may(address(0)), 0);
+
+        vm.expectEmit(true, false, false, false);
+        emit Mate(address(0));
+
+        inputConduit.mate(address(0));
+
+        assertEq(inputConduit.may(address(0)), 1);
+
+        vm.expectEmit(true, false, false, false);
+        emit Hate(address(0));
+
+        inputConduit.hate(address(0));
+
+        assertEq(inputConduit.may(address(0)), 0);
+    }
+
+    function testRevertOnUnauthorizedMethods() public {
+        vm.startPrank(address(0));
+
+        vm.expectRevert("RwaInputConduit3/not-authorized");
+        inputConduit.rely(address(0));
+
+        vm.expectRevert("RwaInputConduit3/not-authorized");
+        inputConduit.deny(address(0));
+
+        vm.expectRevert("RwaInputConduit3/not-authorized");
+        inputConduit.hate(address(0));
+
+        vm.expectRevert("RwaInputConduit3/not-authorized");
+        inputConduit.mate(address(0));
+
+        vm.expectRevert("RwaInputConduit3/not-mate");
+        inputConduit.push();
+
+        vm.expectRevert("RwaInputConduit3/not-mate");
+        inputConduit.push(100);
```

Comment:
Are you sure anything after the first failure is even executing?

Reply:
Yeap

Reply:
Yep, you are correct:
https://book.getfoundry.sh/cheatcodes/expect-revert

> You can also have multiple expectRevert() checks in a single test.
> 
> ```solidity
> function testMultipleExpectReverts() public {
>     vm.expectRevert(abi.encodePacked("INVALID_AMOUNT"));
>     vault.send(user, 0);
> 
>     vm.expectRevert(abi.encodePacked("INVALID_ADDRESS"));
>     vault.send(address(0), 200);
> }
> ```

TIL